File: python/mirage/commands/builtin/redis/mv.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.redis import RedisAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.redis.glob import resolve_glob
from mirage.core.redis.rename import rename
from mirage.core.redis.stat import stat as stat_core
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def _exists(accessor: RedisAccessor, path: str) -> bool:
    try:
        await stat_core(accessor, path)
        return True
    except (FileNotFoundError, ValueError):
        return False


@command("mv", resource="redis", spec=SPECS["mv"], write=True)
async def mv(
    accessor: RedisAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    f: bool = False,
    n: bool = False,
    v: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if accessor.store is None or len(paths) < 2:
        raise ValueError("mv: requires src and dst")
    paths = await resolve_glob(accessor, paths, index)
    if n and await _exists(accessor, paths[1]):
        return None, IOResult()
    await rename(accessor, paths[0], paths[1])
    output = None
    if v:
        output = f"'{paths[0].original}' -> '{paths[1].original}'\n".encode()
    return output, IOResult()

File: typescript/packages/node/src/commands/builtin/redis/mv.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  IOResult,
  ResourceName,
  command,
  mvGeneric,
  specOf,
  type CommandFnResult,
  type CommandOpts,
  type PathSpec,
} from '@struktoai/mirage-core'
import { rename as coreRename } from '../../../core/redis/rename.ts'
import { stat as coreStat } from '../../../core/redis/stat.ts'
import type { RedisAccessor } from '../../../accessor/redis.ts'

function mvCommand(
  accessor: RedisAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length < 2) {
    return Promise.resolve([
      null,
      new IOResult({ exitCode: 1, stderr: new TextEncoder().encode('mv: missing operand\n') }),
    ])
  }
  return mvGeneric(
    paths,
    (src: PathSpec, target: PathSpec) => coreRename(accessor, src, target),
    (p: PathSpec) => coreStat(accessor, p),
    opts.flags.n === true,
    opts.flags.v === true,
    opts.index ?? undefined,
  )
}

export const REDIS_MV = command({
  name: 'mv',
  resource: ResourceName.REDIS,
  spec: specOf('mv'),
  fn: mvCommand,
  write: true,
})
